### PR TITLE
Api root me

### PR DESCRIPTION
**Pull request by Xlitoni**
Here is a first basis for the root-me API.

You should set the API_KEY_ROOTME environment variable. 
The main has become async to show how the API would work.
Only able to query for users by id and challenges by id for now.

### CODE REVIEW

File: src/api/rate_limiter.py
Comment (atxr) [inline src/api/rate_limiter.py:37]: It works but might be overkill
The request already takes some time here, so we may wait too long
What about storing the last request time and add instead of the sleep something like
```py
while datetime.now() - lastRequestTime < 0.050:
    continue
```

**Review by atxr — COMMENTED** [review]
:fire_engine: 
A question here:
What happens if the request fails? Does it pop it from the queue, or try to send it again?

File: src/api/rate_limiter.py
Comment (Xlitoni) [reply src/api/rate_limiter.py:37]: True

Comment (ctmbl) [reply src/api/rate_limiter.py:37]: actually that piece of code is really bad habits, because it not only slows the code it also slows the machine: while a `time.sleep` give the processor back to the kernel this "active waiting" doesn't! 
I would prefer 
```py
delay = datetime.now() - lastRequestTime - 0.050
if delay > 0.001 :
  time.sleep(delay)
  ```

File: README.md
Comment (ctmbl) [inline README.md:17]: I quite don't understand how your README commit ended up in both PRs :laughing: but this is fine I think, you'll probably have a confloicts or a discarded commit once one of the two will be merged though

File: run.sh
Comment (ctmbl) [inline run.sh:12]: I'll look into, this seems weird...

File: src/main.py
Comment (ctmbl) [inline src/main.py:12]: so you choose to use async everywhere? would we just **not** use it instead?

File: src/api/rate_limiter.py
Comment (ctmbl) [inline src/api/rate_limiter.py:30]: You should use `logging` instead :wink:

Comment (ctmbl) [reply src/api/rate_limiter.py:30]: but maybe you're not used to it, i should introduce it in another PR if that's the case

File: src/api/rate_limiter.py
Comment (ctmbl) [inline src/api/rate_limiter.py:31]: I guess thanks to the async this doesn't create an infinite loop?

File: src/api/rootme_api.py
Comment (ctmbl) [inline src/api/rootme_api.py:18]: maybe get this one from env? or at least global constant?

File: src/api/rootme_api.py
Comment (ctmbl) [inline src/api/rootme_api.py:29]: you should use `self.url` here :wink:

File: src/api/rootme_api.py
Comment (ctmbl) [inline src/api/rootme_api.py:40]: You should use `self.url` here too!

**Review by ctmbl — CHANGES_REQUESTED** [review]
Some minor modifications and questions but this looks like a really good starting point!

**Comment by ctmbl** [comment]
> It's a first version, I will indeed update it to make it more robust in case of requests not working. I think trying to send again with a max_retry, and if it still doesn't work, make the requesting function handle the fact that the request does not seem to work.

looks like a good idea :+1:

File: src/main.py
Comment (Xlitoni) [reply src/main.py:12]: well if you use asynchronous function somewhere you should `await` every function that is `async` and a function that awaits another one must be awaited as well. So as soon as you use it the main becomes `async`.

File: src/api/rate_limiter.py
Comment (Xlitoni) [reply src/api/rate_limiter.py:30]: I don't know indeed but i can add it :wink:

File: src/api/rate_limiter.py
Comment (Xlitoni) [reply src/api/rate_limiter.py:31]: Yes this creates an infinite task, but as soon as this asynchronous task is awaiting something either `await self.queue.get()` or the `asyncio.sleep()`, it is not actually being executed.

File: src/api/rootme_api.py
Comment (Xlitoni) [reply src/api/rootme_api.py:18]: Yes indeed :+1:

File: src/api/rootme_api.py
Comment (Xlitoni) [reply src/api/rootme_api.py:29]: Yes :open_mouth:

File: run.sh
Comment (Xlitoni) [reply run.sh:12]: I think it is a problem with my distro, and I found a lot of people were forced to use absolute path. If you find a fix I'm interested tho.

File: src/api/rate_limiter.py
Comment (Xlitoni) [reply src/api/rate_limiter.py:37]: I agree about the time.sleep() but the point of using asyncio.sleep (to be confirmed but that's what I understood) is that it gives this time not to the processor but to other asynchronous tasks that might be waiting somewhere else in the program. So (to me) time.sleep would be fully blocking the program, whereas asyncio.sleep is allowing the rest of the program to still run while this particular task waits.

Comment (ctmbl) [reply src/api/rate_limiter.py:37]: Actually my bad, my point wasn't the `time.sleep` or `asyncio.sleep` it was of not using `while... continue` 
```py
delay = datetime.now() - lastRequestTime - 0.050
if delay > 0.001 :
  asyncio.sleep(delay)
  ```
  is fine to me

File: src/main.py
Comment (ctmbl) [reply src/main.py:12]: ok I get it tks! I'll trust you, does it seem mandatory or could we avoid it entirely?

File: src/api/rate_limiter.py
Comment (ctmbl) [reply src/api/rate_limiter.py:30]: If you're brave enough to dig into it you can, otherwise no worries I'll probably change every print I see when I'll add it :wink:

File: src/api/rate_limiter.py
Comment (ctmbl) [reply src/api/rate_limiter.py:31]: Ok it's great actually! tks for the explanation

**Comment by ctmbl** [comment]
> Well actually that can be tested without retrying, one of the failed request case would be status code 429 - too many requests -> in this case we certainly should not try sending the request again.

I've tested a bit the api too and at first I think taht too but if my memory serves me well someone explained me that this was two entirely different cases... we should test it (force an automatic ban) to be sure

File: src/main.py
Comment (atxr) [reply src/main.py:12]: As soon as you use `await` or `async` in one place, you must use it for all the **callers**.
So I don't think we can nor should avoid it.

File: run.sh
Comment (Xlitoni) [reply run.sh:12]: I'm going to leave it like that since it still works for everyone that way.

File: src/api/rate_limiter.py
Comment (Xlitoni) [reply src/api/rate_limiter.py]: fixed :+1:

**Comment by Xlitoni** [comment]
I think i'll try to add logging to learn it, and in the end i'll try to correct the pylint test :smiling_face_with_tear:

File: src/api/rate_limiter.py
Comment (Xlitoni) [reply src/api/rate_limiter.py]: Now there is logging, a very basic one. I used logging.basicConfig meaning I use the global object logging directly which is considered not a very good habit. But as long as we don't need different loggers this can do the job very simply I think.

**Review by ctmbl — CHANGES_REQUESTED** [review]
Still some changes but that looks gooooooood

File: src/api/rate_limiter.py
Comment (ctmbl) [reply src/api/rate_limiter.py]: [16442e5](https://github.com/iScsc/RootPythia/pull/10/commits/16442e50f3c11c25fbcf298397ecb1694c5beaa8) perfect!

File: src/api/rate_limiter.py
Comment (ctmbl) [reply src/api/rate_limiter.py]: [a86c6d4](https://github.com/iScsc/RootPythia/pull/10/commits/a86c6d4f466cb471916dfd9b0df410f0761f7a9a) :heart:

File: src/api/rate_limiter.py
Comment (ctmbl) [reply src/api/rate_limiter.py]: Actually the static variables seemed fine to me as the `_max_retry` doesn't seems too instance-dependant, but this do the job as well!

Comment (ctmbl) [reply src/api/rate_limiter.py]: @Xlitoni if you wish to let it as is please yourself to resolve this thread yourself and merge the PR I'll approve it :wink: 
Otherwise you can change it to a static attribute that would be cool too!

File: run.sh
Comment (ctmbl) [reply run.sh:18]: still blocking the PR :confused:

**Review by atxr — COMMENTED** [review]
Just one comment here 🏑

File: src/api/rate_limiter.py
Comment (ctmbl) [reply src/api/rate_limiter.py:66]: I thought that too but actually it's constraint by the rootme api itself if I'm not mistaken, there is then no point in making it configurable.

Comment (atxr) [reply src/api/rate_limiter.py:66]: I'll let @Xlitoni close the thread so he can read it

Comment (Xlitoni) [reply src/api/rate_limiter.py:66]: I think we can leave it like that, firstly because I took a pretty generous timeout, and secondly because this is just for style, I don't really see a use-case where we would have to spam the API so much that this limit would actually be reached.

File: run.sh
Comment (Xlitoni) [reply run.sh]: Can just be discarded on merge no ?

Comment (ctmbl) [reply run.sh]: @Xlitoni absolutely but you'll have to solve the conflict first with a `git merge main` :wink:

**Review by ctmbl — APPROVED** [review]
Good to go at last!

**Pull request merged**
